inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/support/AnnotationQueueOverviewDataProvider.java

```diff
@@ -300,60 +350,111 @@ public long getInProgressAmountForDocument(
+
+    public SourceDocument getRandomDocument(
+        AnnotationPageBase aAnnotationPageBase, AnnotationDocument aCurrentAnno)
     {
-        //Create an empty document
-        SourceDocument document = null;
+
+        //First check for a documents that is in Progress for the user, and return this
+        for (int i = 0; i < allAnnotationDocuments.size(); i++)
+        {
+            if (allAnnotationDocuments.get(i).getUser().equals
+                (aAnnotationPageBase.getModelObject().getUser().getUsername())
+                && allAnnotationDocuments.get(i).getState().equals(IN_PROGRESS)
+                && !allAnnotationDocuments.get(i).getName().equals(aCurrentAnno.getName()))
+            {
+                //Found a document in progress
+                for (SourceDocument doc: data)
+                {
+                    if (doc.getName().equals(allAnnotationDocuments.get(i).getName()))
+                    {
+                        return doc;
+                    }
+                }
+            }
+        }
+
+
         Random r = new Random();
+        int attempts = 0;
 
-        while (document == null)
+        //Nothing in Progress, so now return a random document, or create a new annotation document
+        while (true)
         {
-            int i = r.nextInt(data.size() - 1);
-            //If the random chosen document wont surpass the amount of default number combining
-            // "inProgress" for the document + "finished" amount for the document + 1
-            if ((getInProgressAmountForDocument(data.
-                get(i)) +
-                getFinishedAmountForDocument(data.
-                    get(i)))
-                + 1 <= defaultAnnotations)
+            SourceDocument src = data.get(r.nextInt(data.size()));
+
+            for (int i = 0; i < allAnnotationDocuments.size(); i++)
             {
-                //If that was not the case, assign this document
-                document = data.get(r.nextInt(data.size() - 1));
+                if (src.getName().equals(allAnnotationDocuments.get(i).getName()) &&
+                    allAnnotationDocuments.get(i).getUser().equals(
+                        aAnnotationPageBase.getModelObject().getUser().getUsername()))
+                {
+                    if (allAnnotationDocuments.get(i).getState().equals(NEW))
+                    {
+                        return src;
+                    } else {
+                        attempts++;
+                        //Try for 3 times the size of all documents, if still nothing new found
+                        //return null. This is catched and will send the user back to the
+                        //homepage.
+                        if (attempts == data.size() * 3)
+                        {
+                            return null;
+                        }
+                        break;
+                    }
+                }
+
+                if (i == allAnnotationDocuments.size() - 1)
```

Comment:
Please don't create AnnotationDocument instances manually. If you use `listAnnotatableDocuments` and/or tailored queries (as outlined above), this is not necessary. Mind, check the actual code of `listAnnotatableDocuments` and `listAllDocuments`. I could actually imaging that `listAllDocuments` could be improved into a single JOIN query. I'm not suggesting you need to write that though - just thinking.